packages/babel-core/src/transformation/block-hoist-plugin.js

```diff
@@ -41,24 +39,25 @@ const blockHoistPlugin = {
   visitor: {
     Block: {
       exit({ node }) {
-        let hasChange = false;
-        for (let i = 0; i < node.body.length; i++) {
-          const bodyNode = node.body[i];
-          if (bodyNode?._blockHoist != null) {
-            hasChange = true;
-            break;
-          }
-        }
-        if (!hasChange) return;
-
-        node.body = sortBy(node.body, function (bodyNode) {
-          let priority = bodyNode?._blockHoist;
-          if (priority == null) priority = 1;
-          if (priority === true) priority = 2;
+        if (!node.body.find(node => node?._blockHoist != null)) return;
+        // TODO: Babel 9 (?) - remove stabilityMap once Node >= 12 is guaranteed
+        // Array sorting is not stable in earlier Node releases
+        // See: https://v8.dev/blog/array-sort for details
+        const sortIsUnstable = process.versions.node.split(".")[0] < 12 || null;
+        const stabilityMap = new Map(
+          sortIsUnstable && node.body.map((n, idx) => [n, idx]),
+        );
 
-          // Higher priorities should move toward the top.
-          return -1 * priority;
-        });
+        const priority = node => {
+          if (node?._blockHoist == null) return -1;
+          if (node?._blockHoist === true) return -2;
+          return -1 * node?._blockHoist;
+        };
+        node.body.sort(
+          (a, b) =>
+            priority(a) - priority(b) ||
+            stabilityMap.get(a) - stabilityMap.get(b),
+        );
```

Comment:
```suggestion
        node.body.sort(
          (a, b) => {
            const result = priority(a) - priority(b);
            if (!result && sortIsUnstable) {
              return stabilityMap.get(a) - stabilityMap.get(b);
            }
            return result;
        });
```
The existing code adds the overhead to node.js 12+ even if not needed and does a lookup on an empty map (`undefined - undefined` is NaN).

Side note in my example code `new Map` was only run in Node.js < 12, current versions of Node.js can avoid even creating the Map object.

Reply:
Yup, good call.  I was trying to streamline the logic a little, but any elegance in the code isn't worth the overhead of the lookups.  I'll push a change for that shortly.